### DISCUSSION TITLE
What is a Session from end user perspective

### DISCUSSION THREAD

**Discussion by lampajr** [Hyperfoil/Hyperfoil#289] [General]
Hi all,
I was deep diving the Hyperfoil documentation and I was not able to find what exactly a _session_ is. I see that in some cases properly configuring it through `maxSessions` could be required (e.g., in the begginer-guide-2 [1]).

_"Hyperfoil has a fixed limit for concurrency - number of virtual users (sessions) executed in parallel. By default this limit is equal to user arrival rate (usersPerSec), so in this scenario it was 10 concurrent users. However as with all those think-times the session takes several seconds, we will require more than 10 concurrent sessions, even if the virtual users are idle in their think-time. Average session should take 4 seconds pausing plus some time for the requests, so we can expect
little over 40 concurrent users. We’ll add some margin and raise the limit to 60 sessions using the maxSessions property:"_

**Q1:** Does this mean that every virtual user MUST have a free session available to start?

If yes, this would require end users to properly configure that `maxSessions` value which could be quite tricky, mainly for very complex scenarios as it would require to understand which would be the highest number of concurrent virtual users at a given time (is this true?).

**Q2:** Furthermore, IIUC what a _session_ is, I would say it is something much more technical, I mean it is a means through by which Hyperfoil can do what he does. Therefore IMO this should be hidden to the end users as it does not have a real meaning to them, wdyt?

**Q3:** Right now hyperfoil can detect situations where we might have to increase `maxSessions` but it does not provide any possible/reasonable value for it, just `Exceeded session limit`. Is it something that could be improved? Maybe letting Hyperfoil understand which is the proper number of sessions given the benchmark definition (TBH don't if this is really feasible).

I am sorry if I have state some inaccuracies but it is still not fully clear to me.

[1] https://hyperfoil.io/2021/02/09/hf-beginner-guide-2.html

**Comment by franz1981** [comment]
Thanks @lampajr and sorry I've the late reply.
When I have read the very first time about sessions I had your same impression and (still unsolved) questions; what's clear to me is that, as you have rightly pointed, sessions are a technical aspect necessary to Hyperfoil to work but shouldn't be users required to understand till is really unavoidable.
In short, we should provide enough information to not force users to manually set them:  eg a calibration phase that measure the required ones or by an eager estimation using symbolic execution/interpretation of the configured test scenario (no idea if possible).
I believe, after a vote, that would be nice to discuss which feature to propose and maybe add more counters already around which resources are used and how much - to help already users into cutting down resources, while eagerly allocated (that's right now the easiest way to size them without too many run attempts).

**Reply by lampajr** [reply]
Thanks @franz1981, I totally agree with you and I will be happy to discuss about possible improvements on this topic as I think this could save users a lot of time for complex scenarios.
